Comparison against `Class` or `Object` causes panic

Now I found on #285 that:

```ruby
nil == Class #=> works

Class == nil  #=> panic
Object == nil #=> panic
```

This panic occurs if either `Class` or `Object` is the receiver.

```ruby
» Class == nil
Internal Error: runtime error: invalid memory address or nil pointer dereference
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x129a016]

goroutine 1 [running]:
github.com/goby-lang/goby/vm.(*thread).execInstruction.func1(0xc420143020)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/thread.go:71 +0x81
panic(0x1339c20, 0x153d260)
	/usr/local/opt/go/libexec/src/runtime/panic.go:489 +0x2cf
github.com/goby-lang/goby/vm.(*baseObj).Class(0x0, 0x0)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/object.go:39 +0x26
github.com/goby-lang/goby/vm.builtinCommonInstanceMethods.func1.1(0xc420143020, 0xc4201623e0, 0x1, 0x1, 0x0, 0xc4201623e0, 0x0)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/class.go:103 +0x3d
github.com/goby-lang/goby/vm.(*thread).evalBuiltInMethod(0xc420143020, 0x151b9a0, 0xc420081500, 0xc42013afa0, 0x0, 0x1, 0x1, 0x0)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/thread.go:130 +0x202
github.com/goby-lang/goby/vm.glob..func21(0xc420143020, 0xc420160d90, 0xc4201720a0, 0x2, 0x2)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/instruction.go:368 +0x3ea
github.com/goby-lang/goby/vm.(*thread).execInstruction(0xc420143020, 0xc420160d90, 0xc42015e840)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/thread.go:75 +0x96
github.com/goby-lang/goby/vm.(*thread).evalCallFrame(0xc420143020, 0xc420160d90)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/thread.go:42 +0x60
github.com/goby-lang/goby/vm.(*thread).startFromTopFrame(0xc420143020)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/thread.go:36 +0x55
github.com/goby-lang/goby/vm.(*VM).startFromTopFrame(0xc420024870)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/vm.go:190 +0x2f
github.com/goby-lang/goby/vm.(*VM).REPLExec(0xc420024870, 0xc420176008, 0x1, 0x1)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/vm/repl.go:42 +0x64e
github.com/goby-lang/goby/igb.StartIgb(0x139e000, 0x5)
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/igb/repl.go:215 +0x152f
main.main()
	/Users/hachi8833/deve/golang/gopath/sys/src/github.com/goby-lang/goby/goby.go:27 +0x657
```

This should be fixed after #285 has been merged.

--- a/vm/class.go
+++ b/vm/class.go
@@ -505,10 +505,12 @@ func builtinCommonInstanceMethods() []*BuiltInMethodObject {
 						if receiverClass.Name == gobyClass.Name {
 							return TRUE
 						}
-						receiverClass = receiverClass.superClass
-						if receiverClass == nil {
+
+						if receiverClass.Name == objectClass {
 							break
 						}
+
+						receiverClass = receiverClass.superClass
 					}
 					return FALSE
 				}

--- a/vm/class.go
+++ b/vm/class.go
@@ -87,16 +87,10 @@ func initClassClass() *RClass {
 }
 
 func (c *RClass) inherits(sc *RClass) {
-	if c.superClass == c.pseudoSuperClass {
-		c.superClass = sc
-	}
+	c.superClass = sc
 	c.pseudoSuperClass = sc
-
-	if sc.singletonClass == nil {
-		panic("!!!!!!!!!!!")
-	}
-
-	c.singletonClass = sc.singletonClass
+	c.singletonClass.superClass = sc.singletonClass
+	c.singletonClass.pseudoSuperClass = sc.singletonClass
 }
 
 func initObjectClass(c *RClass) *RClass {
@@ -117,10 +111,12 @@ func initObjectClass(c *RClass) *RClass {
 		isModule:     false,
 		baseObj:      &baseObj{class: c, InstanceVariables: newEnvironment()},
 		Singleton:    true,
+		superClass:   c,
 	}
 
 	objectClass.singletonClass = singletonClass
-	objectClass.inherits(objectClass)
+	objectClass.superClass = objectClass
+	objectClass.pseudoSuperClass = objectClass
 	c.inherits(objectClass)
 
 	objectClass.setBuiltInMethods(builtinClassClassMethods(), true)
@@ -131,6 +127,14 @@ func initObjectClass(c *RClass) *RClass {
 
 func builtinCommonInstanceMethods() []*BuiltInMethodObject {
 	return []*BuiltInMethodObject{
+		{
+			Name: "singleton_class",
+			Fn: func(receiver Object) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *callFrame) Object {
+					return receiver.SingletonClass()
+				}
+			},
+		},
 		{
 			// General method for comparing equalty of the objects
 			//
@@ -923,7 +927,8 @@ func (vm *VM) initializeClass(name string, isModule bool) *RClass {
 	class.isModule = isModule
 	singletonClass := vm.createRClass(fmt.Sprintf("#<Class:%s>", name))
 	singletonClass.Singleton = true
-	class.singletonClass = class
+	class.singletonClass = singletonClass
+	class.inherits(vm.objectClass)
 
 	return class
 }
@@ -949,10 +954,7 @@ func (vm *VM) createRClass(className string) *RClass {
 // toString returns the basic inspected result (which is class name) of current class
 // TODO: Singleton class's inspect() should also mark if it's a singleton class explicitly.
 func (c *RClass) toString() string {
-	if c.isModule {
-		return "<Module:" + c.Name + ">"
-	}
-	return "<Class:" + c.Name + ">"
+	return c.Name
 }
 
 func (c *RClass) toJSON() string {
@@ -973,9 +975,7 @@ func (c *RClass) setBuiltInMethods(methodList []*BuiltInMethodObject, classMetho
 
 func (c *RClass) lookupClassMethod(methodName string) Object {
 	method, ok := c.ClassMethods.get(methodName)
-	//
-	//fmt.Println(c.Name)
-	//fmt.Println(methodName)
+
 	if !ok {
 		if c.superClass != nil && c.superClass != c {
 			//fmt.Printf("Finding class method: %s on %s. Superclass is %s\n", methodName, c.Name, c.superClass.Name)
@@ -997,15 +997,15 @@ func (c *RClass) lookupInstanceMethod(methodName string) Object {
 	if !ok {
 		if c.superClass != nil && c.superClass != c {
 			//fmt.Printf("Finding instance method: %s on %s. Superclass is %s\n", methodName, c.Name, c.superClass.Name)
+			if c.Name == classClass {
+				return nil
+			}
+
 			return c.superClass.lookupInstanceMethod(methodName)
 		}
 
 		if c.singletonClass != nil {
 			return c.singletonClass.lookupInstanceMethod(methodName)
-		}
-
-		if c.class != nil && c.Name != objectClass {
-			return c.class.lookupInstanceMethod(methodName)
 		}
 
 		return nil
@@ -1021,10 +1021,7 @@ func (c *RClass) lookupConstant(constName string, findInScope bool) *Pointer {
 		if findInScope && c.scope != nil {
 			return c.scope.lookupConstant(constName, true)
 		}
-		//
-		//fmt.Println(constName)
-		//fmt.Println(c.Name)
-		//fmt.Println(c.superClass.Name)
+
 		if c.superClass != nil && c.Name != objectClass {
 			return c.superClass.lookupConstant(constName, false)
 		}

--- a/vm/class.go
+++ b/vm/class.go
@@ -123,6 +123,7 @@ func initObjectClass(c *RClass) *RClass {
 	objectClass.inherits(objectClass)
 	c.inherits(objectClass)
 
+	objectClass.setBuiltInMethods(builtinClassClassMethods(), true)
 	objectClass.setBuiltInMethods(builtinCommonInstanceMethods(), false)
 
 	return objectClass
@@ -972,9 +973,12 @@ func (c *RClass) setBuiltInMethods(methodList []*BuiltInMethodObject, classMetho
 
 func (c *RClass) lookupClassMethod(methodName string) Object {
 	method, ok := c.ClassMethods.get(methodName)
-
+	//
+	//fmt.Println(c.Name)
+	//fmt.Println(methodName)
 	if !ok {
-		if c.superClass != nil {
+		if c.superClass != nil && c.superClass != c {
+			//fmt.Printf("Finding class method: %s on %s. Superclass is %s\n", methodName, c.Name, c.superClass.Name)
 			return c.superClass.lookupClassMethod(methodName)
 		}
 		if c.class != nil {
@@ -989,12 +993,18 @@ func (c *RClass) lookupClassMethod(methodName string) Object {
 func (c *RClass) lookupInstanceMethod(methodName string) Object {
 	method, ok := c.Methods.get(methodName)
 
+	//fmt.Println(c.Name)
 	if !ok {
-		if c.superClass != nil {
+		if c.superClass != nil && c.superClass != c {
+			//fmt.Printf("Finding instance method: %s on %s. Superclass is %s\n", methodName, c.Name, c.superClass.Name)
 			return c.superClass.lookupInstanceMethod(methodName)
 		}
 
-		if c.class != nil {
+		if c.singletonClass != nil {
+			return c.singletonClass.lookupInstanceMethod(methodName)
+		}
+
+		if c.class != nil && c.Name != objectClass {
 			return c.class.lookupInstanceMethod(methodName)
 		}
 
@@ -1011,10 +1021,10 @@ func (c *RClass) lookupConstant(constName string, findInScope bool) *Pointer {
 		if findInScope && c.scope != nil {
 			return c.scope.lookupConstant(constName, true)
 		}
-
-		fmt.Println(constName)
-		fmt.Println(c.Name)
-		fmt.Println(c.superClass.Name)
+		//
+		//fmt.Println(constName)
+		//fmt.Println(c.Name)
+		//fmt.Println(c.superClass.Name)
 		if c.superClass != nil && c.Name != objectClass {
 			return c.superClass.lookupConstant(constName, false)
 		}

--- a/vm/class.go
+++ b/vm/class.go
@@ -63,13 +63,40 @@ func initClassClass() *RClass {
 		Methods:      newEnvironment(),
 		ClassMethods: newEnvironment(),
 		constants:    make(map[string]*Pointer),
+		baseObj:      &baseObj{},
 	}
+
+	singletonClass := &RClass{
+		Name:         "#<Class:Class>",
+		Methods:      newEnvironment(),
+		ClassMethods: newEnvironment(),
+		constants:    make(map[string]*Pointer),
+		isModule:     false,
+		baseObj:      &baseObj{class: classClass, InstanceVariables: newEnvironment()},
+		Singleton:    true,
+	}
+
+	classClass.class = classClass
+	classClass.singletonClass = singletonClass
 
 	classClass.setBuiltInMethods(builtinCommonInstanceMethods(), false)
 	classClass.setBuiltInMethods(builtinCommonInstanceMethods(), true)
 	classClass.setBuiltInMethods(builtinClassClassMethods(), true)
 
 	return classClass
+}
+
+func (c *RClass) inherits(sc *RClass) {
+	if c.superClass == c.pseudoSuperClass {
+		c.superClass = sc
+	}
+	c.pseudoSuperClass = sc
+
+	if sc.singletonClass == nil {
+		panic("!!!!!!!!!!!")
+	}
+
+	c.singletonClass = sc.singletonClass
 }
 
 func initObjectClass(c *RClass) *RClass {
@@ -79,7 +106,22 @@ func initObjectClass(c *RClass) *RClass {
 		ClassMethods: newEnvironment(),
 		Methods:      newEnvironment(),
 		constants:    make(map[string]*Pointer),
+		baseObj:      &baseObj{class: c},
 	}
+
+	singletonClass := &RClass{
+		Name:         "#<Class:Object>",
+		Methods:      newEnvironment(),
+		ClassMethods: newEnvironment(),
+		constants:    make(map[string]*Pointer),
+		isModule:     false,
+		baseObj:      &baseObj{class: c, InstanceVariables: newEnvironment()},
+		Singleton:    true,
+	}
+
+	objectClass.singletonClass = singletonClass
+	objectClass.inherits(objectClass)
+	c.inherits(objectClass)
 
 	objectClass.setBuiltInMethods(builtinCommonInstanceMethods(), false)
 
@@ -878,6 +920,9 @@ func builtinClassClassMethods() []*BuiltInMethodObject {
 func (vm *VM) initializeClass(name string, isModule bool) *RClass {
 	class := vm.createRClass(name)
 	class.isModule = isModule
+	singletonClass := vm.createRClass(fmt.Sprintf("#<Class:%s>", name))
+	singletonClass.Singleton = true
+	class.singletonClass = class
 
 	return class
 }
@@ -967,7 +1012,10 @@ func (c *RClass) lookupConstant(constName string, findInScope bool) *Pointer {
 			return c.scope.lookupConstant(constName, true)
 		}
 
-		if c.superClass != nil {
+		fmt.Println(constName)
+		fmt.Println(c.Name)
+		fmt.Println(c.superClass.Name)
+		if c.superClass != nil && c.Name != objectClass {
 			return c.superClass.lookupConstant(constName, false)
 		}
 

--- a/vm/class.go
+++ b/vm/class.go
@@ -57,10 +57,10 @@ type RClass struct {
 
 func initClassClass() *RClass {
 	classClass := &RClass{
-		Name:         classClass,
-		Methods:      newEnvironment(),
-		constants:    make(map[string]*Pointer),
-		baseObj:      &baseObj{},
+		Name:      classClass,
+		Methods:   newEnvironment(),
+		constants: make(map[string]*Pointer),
+		baseObj:   &baseObj{},
 	}
 
 	singletonClass := &RClass{
@@ -89,11 +89,11 @@ func (c *RClass) inherits(sc *RClass) {
 
 func initObjectClass(c *RClass) *RClass {
 	objectClass := &RClass{
-		Name:         objectClass,
-		class:        c,
-		Methods:      newEnvironment(),
-		constants:    make(map[string]*Pointer),
-		baseObj:      &baseObj{class: c},
+		Name:      objectClass,
+		class:     c,
+		Methods:   newEnvironment(),
+		constants: make(map[string]*Pointer),
+		baseObj:   &baseObj{class: c},
 	}
 
 	singletonClass := &RClass{
@@ -969,10 +969,8 @@ func (c *RClass) setBuiltInMethods(methodList []*BuiltInMethodObject, classMetho
 func (c *RClass) lookupInstanceMethod(methodName string) Object {
 	method, ok := c.Methods.get(methodName)
 
-	//fmt.Println(c.Name)
 	if !ok {
 		if c.superClass != nil && c.superClass != c {
-			//fmt.Printf("Finding instance method: %s on %s. Superclass is %s\n", methodName, c.Name, c.superClass.Name)
 			if c.Name == classClass {
 				return nil
 			}

--- a/vm/class.go
+++ b/vm/class.go
@@ -75,8 +75,6 @@ func initClassClass() *RClass {
 	classClass.class = classClass
 	classClass.singletonClass = singletonClass
 
-	classClass.setBuiltInMethods(builtinCommonInstanceMethods(), false)
-	classClass.setBuiltInMethods(builtinCommonInstanceMethods(), true)
 	classClass.setBuiltInMethods(builtinClassClassMethods(), true)
 
 	return classClass
@@ -113,7 +111,7 @@ func initObjectClass(c *RClass) *RClass {
 	objectClass.pseudoSuperClass = objectClass
 	c.inherits(objectClass)
 
-	objectClass.setBuiltInMethods(builtinClassClassMethods(), true)
+	objectClass.setBuiltInMethods(builtinCommonInstanceMethods(), true)
 	objectClass.setBuiltInMethods(builtinCommonInstanceMethods(), false)
 
 	return objectClass

--- a/vm/class.go
+++ b/vm/class.go
@@ -968,26 +968,9 @@ func (c *RClass) setBuiltInMethods(methodList []*BuiltInMethodObject, classMetho
 
 	if classMethods {
 		for _, m := range methodList {
-			c.ClassMethods.set(m.Name, m)
+			c.singletonClass.Methods.set(m.Name, m)
 		}
 	}
-}
-
-func (c *RClass) lookupClassMethod(methodName string) Object {
-	method, ok := c.ClassMethods.get(methodName)
-
-	if !ok {
-		if c.superClass != nil && c.superClass != c {
-			//fmt.Printf("Finding class method: %s on %s. Superclass is %s\n", methodName, c.Name, c.superClass.Name)
-			return c.superClass.lookupClassMethod(methodName)
-		}
-		if c.class != nil {
-			return c.class.lookupClassMethod(methodName)
-		}
-		return nil
-	}
-
-	return method
 }
 
 func (c *RClass) lookupInstanceMethod(methodName string) Object {
@@ -1002,10 +985,6 @@ func (c *RClass) lookupInstanceMethod(methodName string) Object {
 			}
 
 			return c.superClass.lookupInstanceMethod(methodName)
-		}
-
-		if c.singletonClass != nil {
-			return c.singletonClass.lookupInstanceMethod(methodName)
 		}
 
 		return nil

--- a/vm/class.go
+++ b/vm/class.go
@@ -47,11 +47,11 @@ type RClass struct {
 	superClass *RClass
 	// Class points to this class's class, which should be ClassClass
 	class *RClass
-	// Singleton is a flag marks if this class a singleton class
-	Singleton bool
-	isModule  bool
-	constants map[string]*Pointer
-	scope     *RClass
+	// isSingleton is a flag marks if this class a singleton class
+	isSingleton bool
+	isModule    bool
+	constants   map[string]*Pointer
+	scope       *RClass
 	*baseObj
 }
 
@@ -64,12 +64,12 @@ func initClassClass() *RClass {
 	}
 
 	singletonClass := &RClass{
-		Name:         "#<Class:Class>",
-		Methods:      newEnvironment(),
-		constants:    make(map[string]*Pointer),
-		isModule:     false,
-		baseObj:      &baseObj{class: classClass, InstanceVariables: newEnvironment()},
-		Singleton:    true,
+		Name:        "#<Class:Class>",
+		Methods:     newEnvironment(),
+		constants:   make(map[string]*Pointer),
+		isModule:    false,
+		baseObj:     &baseObj{class: classClass, InstanceVariables: newEnvironment()},
+		isSingleton: true,
 	}
 
 	classClass.class = classClass
@@ -97,13 +97,13 @@ func initObjectClass(c *RClass) *RClass {
 	}
 
 	singletonClass := &RClass{
-		Name:         "#<Class:Object>",
-		Methods:      newEnvironment(),
-		constants:    make(map[string]*Pointer),
-		isModule:     false,
-		baseObj:      &baseObj{class: c, InstanceVariables: newEnvironment()},
-		Singleton:    true,
-		superClass:   c,
+		Name:        "#<Class:Object>",
+		Methods:     newEnvironment(),
+		constants:   make(map[string]*Pointer),
+		isModule:    false,
+		baseObj:     &baseObj{class: c, InstanceVariables: newEnvironment()},
+		isSingleton: true,
+		superClass:  c,
 	}
 
 	objectClass.singletonClass = singletonClass
@@ -920,7 +920,7 @@ func (vm *VM) initializeClass(name string, isModule bool) *RClass {
 	class := vm.createRClass(name)
 	class.isModule = isModule
 	singletonClass := vm.createRClass(fmt.Sprintf("#<Class:%s>", name))
-	singletonClass.Singleton = true
+	singletonClass.isSingleton = true
 	class.singletonClass = singletonClass
 	class.inherits(vm.objectClass)
 

--- a/vm/class.go
+++ b/vm/class.go
@@ -40,8 +40,6 @@ type RClass struct {
 	Name string
 	// Methods contains its instances' methods
 	Methods *environment
-	// ClassMethods contains this class's methods
-	ClassMethods *environment
 	// pseudoSuperClass points to the class it inherits
 	pseudoSuperClass *RClass
 	// This is the class where we should looking for a method.
@@ -61,7 +59,6 @@ func initClassClass() *RClass {
 	classClass := &RClass{
 		Name:         classClass,
 		Methods:      newEnvironment(),
-		ClassMethods: newEnvironment(),
 		constants:    make(map[string]*Pointer),
 		baseObj:      &baseObj{},
 	}
@@ -69,7 +66,6 @@ func initClassClass() *RClass {
 	singletonClass := &RClass{
 		Name:         "#<Class:Class>",
 		Methods:      newEnvironment(),
-		ClassMethods: newEnvironment(),
 		constants:    make(map[string]*Pointer),
 		isModule:     false,
 		baseObj:      &baseObj{class: classClass, InstanceVariables: newEnvironment()},
@@ -97,7 +93,6 @@ func initObjectClass(c *RClass) *RClass {
 	objectClass := &RClass{
 		Name:         objectClass,
 		class:        c,
-		ClassMethods: newEnvironment(),
 		Methods:      newEnvironment(),
 		constants:    make(map[string]*Pointer),
 		baseObj:      &baseObj{class: c},
@@ -106,7 +101,6 @@ func initObjectClass(c *RClass) *RClass {
 	singletonClass := &RClass{
 		Name:         "#<Class:Object>",
 		Methods:      newEnvironment(),
-		ClassMethods: newEnvironment(),
 		constants:    make(map[string]*Pointer),
 		isModule:     false,
 		baseObj:      &baseObj{class: c, InstanceVariables: newEnvironment()},
@@ -940,7 +934,6 @@ func (vm *VM) createRClass(className string) *RClass {
 	return &RClass{
 		Name:             className,
 		Methods:          newEnvironment(),
-		ClassMethods:     newEnvironment(),
 		pseudoSuperClass: objectClass,
 		superClass:       objectClass,
 		constants:        make(map[string]*Pointer),

--- a/vm/class_test.go
+++ b/vm/class_test.go
@@ -539,51 +539,52 @@ func TestGeneralIsNilMethodFail(t *testing.T) {
 	}
 }
 
-func TestGeneralIsAMethod(t *testing.T) {
-	tests := []struct {
-		input    string
-		expected bool
-	}{
-		{`123.is_a(Integer)`, true},
-		{`123.is_a(Object)`, true},
-		{`123.is_a(String)`, false},
-		{`123.is_a(Range)`, false},
-		{`"Hello World".is_a(String)`, true},
-		{`"Hello World".is_a(Object)`, true},
-		{`"Hello World".is_a(Boolean)`, false},
-		{`"Hello World".is_a(Array)`, false},
-		{`(2..10).is_a(Range)`, true},
-		{`(2..10).is_a(Object)`, true},
-		{`(2..10).is_a(Null)`, false},
-		{`(2..10).is_a(Hash)`, false},
-		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Hash)`, true},
-		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Object)`, true},
-		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Class)`, false},
-		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Array)`, false},
-		{`[1, 2, 3, 4, 5].is_a(Array)`, true},
-		{`[1, 2, 3, 4, 5].is_a(Object)`, true},
-		{`[1, 2, 3, 4, 5].is_a(Null)`, false},
-		{`[1, 2, 3, 4, 5].is_a(String)`, false},
-		{`true.is_a(Boolean)`, true},
-		{`true.is_a(Object)`, true},
-		{`true.is_a(Array)`, false},
-		{`true.is_a(Integer)`, false},
-		{`String.is_a(Class)`, true},
-		{`String.is_a(String)`, false},
-		{`String.is_a(Array)`, false},
-		{`nil.is_a(Null)`, true},
-		{`nil.is_a(Object)`, true},
-		{`nil.is_a(String)`, false},
-		{`nil.is_a(Range)`, false},
-	}
 
-	for i, tt := range tests {
-		vm := initTestVM()
-		evaluated := vm.testEval(t, tt.input)
-		checkExpected(t, i, evaluated, tt.expected)
-		vm.checkCFP(t, i, 0)
-	}
-}
+//func TestGeneralIsAMethod(t *testing.T) {
+//	tests := []struct {
+//		input    string
+//		expected bool
+//	}{
+//		{`123.is_a(Integer)`, true},
+//		{`123.is_a(Object)`, true},
+//		{`123.is_a(String)`, false},
+//		{`123.is_a(Range)`, false},
+//		{`"Hello World".is_a(String)`, true},
+//		{`"Hello World".is_a(Object)`, true},
+//		{`"Hello World".is_a(Boolean)`, false},
+//		{`"Hello World".is_a(Array)`, false},
+//		{`(2..10).is_a(Range)`, true},
+//		{`(2..10).is_a(Object)`, true},
+//		{`(2..10).is_a(Null)`, false},
+//		{`(2..10).is_a(Hash)`, false},
+//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Hash)`, true},
+//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Object)`, true},
+//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Class)`, false},
+//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Array)`, false},
+//		{`[1, 2, 3, 4, 5].is_a(Array)`, true},
+//		{`[1, 2, 3, 4, 5].is_a(Object)`, true},
+//		{`[1, 2, 3, 4, 5].is_a(Null)`, false},
+//		{`[1, 2, 3, 4, 5].is_a(String)`, false},
+//		{`true.is_a(Boolean)`, true},
+//		{`true.is_a(Object)`, true},
+//		{`true.is_a(Array)`, false},
+//		{`true.is_a(Integer)`, false},
+//		{`String.is_a(Class)`, true},
+//		{`String.is_a(String)`, false},
+//		{`String.is_a(Array)`, false},
+//		{`nil.is_a(Null)`, true},
+//		{`nil.is_a(Object)`, true},
+//		{`nil.is_a(String)`, false},
+//		{`nil.is_a(Range)`, false},
+//	}
+//
+//	for i, tt := range tests {
+//		vm := initTestVM()
+//		evaluated := vm.testEval(t, tt.input)
+//		checkExpected(t, i, evaluated, tt.expected)
+//		vm.checkCFP(t, i, 0)
+//	}
+//}
 
 func TestClassGeneralComparisonOperation(t *testing.T) {
 	tests := []struct {
@@ -703,8 +704,15 @@ func TestClassSuperclassClassMethod(t *testing.T) {
 		{`Range.superclass.name`, "Object"},
 		{`Hash.superclass.name`, "Object"},
 		{`Array.superclass.name`, "Object"},
-		{`Object.superclass`, nil},
-		{`Class.superclass`, nil},
+		{`Object.superclass.name`, "Object"},
+		{`Class.superclass.name`, "Object"},
+		{`
+		module Bar; end
+		class Foo
+		  include Bar
+		end
+		Foo.superclass.name
+		`, "Object"},
 	}
 
 	for i, tt := range tests {

--- a/vm/class_test.go
+++ b/vm/class_test.go
@@ -678,9 +678,9 @@ func TestClassNameClassMethodFail(t *testing.T) {
 		errType string
 		errMsg  string
 	}{
-		{`"Taipei".name`, UndefinedMethodError, "UndefinedMethodError: Undefined Method '#name' for Taipei"},
-		{`123.name`, UndefinedMethodError, "UndefinedMethodError: Undefined Method '#name' for 123"},
-		{`true.name`, UndefinedMethodError, "UndefinedMethodError: Undefined Method '#name' for true"},
+		{`"Taipei".name`, UndefinedMethodError, "UndefinedMethodError: Undefined Method 'name' for Taipei"},
+		{`123.name`, UndefinedMethodError, "UndefinedMethodError: Undefined Method 'name' for 123"},
+		{`true.name`, UndefinedMethodError, "UndefinedMethodError: Undefined Method 'name' for true"},
 		{`Integer.name(Integer)`, ArgumentError, "ArgumentError: Expect 0 argument. got: 1"},
 		{`String.name(Hash, Array)`, ArgumentError, "ArgumentError: Expect 0 argument. got: 2"},
 	}
@@ -723,6 +723,27 @@ func TestClassSuperclassClassMethod(t *testing.T) {
 	}
 }
 
+func TestClassSuperclassClassMethodFail(t *testing.T) {
+	testsFail := []struct {
+		input   string
+		errType string
+		errMsg  string
+	}{
+		{`"Taipei".superclass`, UndefinedMethodError, "UndefinedMethodError: Undefined Method 'superclass' for Taipei"},
+		{`123.superclass`, UndefinedMethodError, "UndefinedMethodError: Undefined Method 'superclass' for 123"},
+		{`true.superclass`, UndefinedMethodError, "UndefinedMethodError: Undefined Method 'superclass' for true"},
+		{`Integer.superclass(Integer)`, ArgumentError, "ArgumentError: Expect 0 argument. got: 1"},
+		{`String.superclass(Hash, Array)`, ArgumentError, "ArgumentError: Expect 0 argument. got: 2"},
+	}
+
+	for i, tt := range testsFail {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input)
+		checkError(t, i, evaluated, tt.errType, tt.errMsg)
+		vm.checkCFP(t, i, 1)
+	}
+}
+
 func TestClassSingletonClassClassMethod(t *testing.T) {
 	tests := []struct {
 		input    string
@@ -746,26 +767,5 @@ func TestClassSingletonClassClassMethod(t *testing.T) {
 		evaluated := vm.testEval(t, tt.input)
 		checkExpected(t, i, evaluated, tt.expected)
 		vm.checkCFP(t, i, 0)
-	}
-}
-
-func TestClassSuperclassClassMethodFail(t *testing.T) {
-	testsFail := []struct {
-		input   string
-		errType string
-		errMsg  string
-	}{
-		{`"Taipei".superclass`, UndefinedMethodError, "UndefinedMethodError: Undefined Method '#superclass' for Taipei"},
-		{`123.superclass`, UndefinedMethodError, "UndefinedMethodError: Undefined Method '#superclass' for 123"},
-		{`true.superclass`, UndefinedMethodError, "UndefinedMethodError: Undefined Method '#superclass' for true"},
-		{`Integer.superclass(Integer)`, ArgumentError, "ArgumentError: Expect 0 argument. got: 1"},
-		{`String.superclass(Hash, Array)`, ArgumentError, "ArgumentError: Expect 0 argument. got: 2"},
-	}
-
-	for i, tt := range testsFail {
-		vm := initTestVM()
-		evaluated := vm.testEval(t, tt.input)
-		checkError(t, i, evaluated, tt.errType, tt.errMsg)
-		vm.checkCFP(t, i, 1)
 	}
 }

--- a/vm/class_test.go
+++ b/vm/class_test.go
@@ -723,6 +723,32 @@ func TestClassSuperclassClassMethod(t *testing.T) {
 	}
 }
 
+func TestClassSingletonClassClassMethod(t *testing.T) {
+	tests := []struct {
+		input    string
+		expected interface{}
+	}{
+		{`Integer.singleton_class.name`, "#<Class:Integer>"},
+		{`Integer.singleton_class.superclass.name`, "#<Class:Object>"},
+		{`
+		class Bar; end
+		Bar.singleton_class.name
+		`, "#<Class:Bar>"},
+		{`
+		class Bar; end
+		class Foo < Bar; end
+		Foo.singleton_class.superclass.name
+		`, "#<Class:Bar>"},
+	}
+
+	for i, tt := range tests {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input)
+		checkExpected(t, i, evaluated, tt.expected)
+		vm.checkCFP(t, i, 0)
+	}
+}
+
 func TestClassSuperclassClassMethodFail(t *testing.T) {
 	testsFail := []struct {
 		input   string

--- a/vm/class_test.go
+++ b/vm/class_test.go
@@ -600,10 +600,9 @@ func TestClassGeneralComparisonOperation(t *testing.T) {
 		{`Integer == Integer`, true},
 		{`Integer == String`, false},
 		{`123.class == Integer`, true},
-		// TODO: Issue #289 Comparing to Object cause panic
-		//{`Integer == Object`, false},
-		//{`Integer.superclass == Object`, true},
-		//{`123.class.superclass == Object`, true},
+		{`Integer == Object`, false},
+		{`Integer.superclass == Object`, true},
+		{`123.class.superclass == Object`, true},
 		{`Integer != 123`, true},
 		{`Integer != "123"`, true},
 		{`Integer != "124"`, true},
@@ -613,10 +612,9 @@ func TestClassGeneralComparisonOperation(t *testing.T) {
 		{`Integer != Integer`, false},
 		{`Integer != String`, true},
 		{`123.class != Integer`, false},
-		// TODO: Issue #289 Comparing to Object cause panic
-		//{`Integer != Object`, true},
-		//{`Integer.superclass != Object`, false},
-		//{`123.class.superclass != Object`, false},
+		{`Integer != Object`, true},
+		{`Integer.superclass != Object`, false},
+		{`123.class.superclass != Object`, false},
 	}
 
 	for i, tt := range tests {

--- a/vm/class_test.go
+++ b/vm/class_test.go
@@ -539,53 +539,6 @@ func TestGeneralIsNilMethodFail(t *testing.T) {
 	}
 }
 
-
-//func TestGeneralIsAMethod(t *testing.T) {
-//	tests := []struct {
-//		input    string
-//		expected bool
-//	}{
-//		{`123.is_a(Integer)`, true},
-//		{`123.is_a(Object)`, true},
-//		{`123.is_a(String)`, false},
-//		{`123.is_a(Range)`, false},
-//		{`"Hello World".is_a(String)`, true},
-//		{`"Hello World".is_a(Object)`, true},
-//		{`"Hello World".is_a(Boolean)`, false},
-//		{`"Hello World".is_a(Array)`, false},
-//		{`(2..10).is_a(Range)`, true},
-//		{`(2..10).is_a(Object)`, true},
-//		{`(2..10).is_a(Null)`, false},
-//		{`(2..10).is_a(Hash)`, false},
-//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Hash)`, true},
-//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Object)`, true},
-//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Class)`, false},
-//		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Array)`, false},
-//		{`[1, 2, 3, 4, 5].is_a(Array)`, true},
-//		{`[1, 2, 3, 4, 5].is_a(Object)`, true},
-//		{`[1, 2, 3, 4, 5].is_a(Null)`, false},
-//		{`[1, 2, 3, 4, 5].is_a(String)`, false},
-//		{`true.is_a(Boolean)`, true},
-//		{`true.is_a(Object)`, true},
-//		{`true.is_a(Array)`, false},
-//		{`true.is_a(Integer)`, false},
-//		{`String.is_a(Class)`, true},
-//		{`String.is_a(String)`, false},
-//		{`String.is_a(Array)`, false},
-//		{`nil.is_a(Null)`, true},
-//		{`nil.is_a(Object)`, true},
-//		{`nil.is_a(String)`, false},
-//		{`nil.is_a(Range)`, false},
-//	}
-//
-//	for i, tt := range tests {
-//		vm := initTestVM()
-//		evaluated := vm.testEval(t, tt.input)
-//		checkExpected(t, i, evaluated, tt.expected)
-//		vm.checkCFP(t, i, 0)
-//	}
-//}
-
 func TestClassGeneralComparisonOperation(t *testing.T) {
 	tests := []struct {
 		input    string
@@ -615,6 +568,52 @@ func TestClassGeneralComparisonOperation(t *testing.T) {
 		{`Integer != Object`, true},
 		{`Integer.superclass != Object`, false},
 		{`123.class.superclass != Object`, false},
+	}
+
+	for i, tt := range tests {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input)
+		checkExpected(t, i, evaluated, tt.expected)
+		vm.checkCFP(t, i, 0)
+	}
+}
+
+func TestGeneralIsAMethod(t *testing.T) {
+	tests := []struct {
+		input    string
+		expected bool
+	}{
+		{`123.is_a(Integer)`, true},
+		{`123.is_a(Object)`, true},
+		{`123.is_a(String)`, false},
+		{`123.is_a(Range)`, false},
+		{`"Hello World".is_a(String)`, true},
+		{`"Hello World".is_a(Object)`, true},
+		{`"Hello World".is_a(Boolean)`, false},
+		{`"Hello World".is_a(Array)`, false},
+		{`(2..10).is_a(Range)`, true},
+		{`(2..10).is_a(Object)`, true},
+		{`(2..10).is_a(Null)`, false},
+		{`(2..10).is_a(Hash)`, false},
+		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Hash)`, true},
+		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Object)`, true},
+		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Class)`, false},
+		{`{ a: 1, b: "2", c: ["Goby", 123] }.is_a(Array)`, false},
+		{`[1, 2, 3, 4, 5].is_a(Array)`, true},
+		{`[1, 2, 3, 4, 5].is_a(Object)`, true},
+		{`[1, 2, 3, 4, 5].is_a(Null)`, false},
+		{`[1, 2, 3, 4, 5].is_a(String)`, false},
+		{`true.is_a(Boolean)`, true},
+		{`true.is_a(Object)`, true},
+		{`true.is_a(Array)`, false},
+		{`true.is_a(Integer)`, false},
+		{`String.is_a(Class)`, true},
+		{`String.is_a(String)`, false},
+		{`String.is_a(Array)`, false},
+		{`nil.is_a(Null)`, true},
+		{`nil.is_a(Object)`, true},
+		{`nil.is_a(String)`, false},
+		{`nil.is_a(Range)`, false},
 	}
 
 	for i, tt := range tests {

--- a/vm/error_test.go
+++ b/vm/error_test.go
@@ -47,12 +47,12 @@ func TestUnsupportedMethodError(t *testing.T) {
 		input    string
 		errorMsg string
 	}{
-		{`String.new`, "UnsupportedMethodError: Unsupported Method #new for <Class:String>"},
-		{`Integer.new`, "UnsupportedMethodError: Unsupported Method #new for <Class:Integer>"},
-		{`Hash.new`, "UnsupportedMethodError: Unsupported Method #new for <Class:Hash>"},
-		{`Array.new`, "UnsupportedMethodError: Unsupported Method #new for <Class:Array>"},
-		{`Boolean.new`, "UnsupportedMethodError: Unsupported Method #new for <Class:Boolean>"},
-		{`Null.new`, "UnsupportedMethodError: Unsupported Method #new for <Class:Null>"},
+		{`String.new`, "UnsupportedMethodError: Unsupported Method #new for String"},
+		{`Integer.new`, "UnsupportedMethodError: Unsupported Method #new for Integer"},
+		{`Hash.new`, "UnsupportedMethodError: Unsupported Method #new for Hash"},
+		{`Array.new`, "UnsupportedMethodError: Unsupported Method #new for Array"},
+		{`Boolean.new`, "UnsupportedMethodError: Unsupported Method #new for Boolean"},
+		{`Null.new`, "UnsupportedMethodError: Unsupported Method #new for Null"},
 	}
 
 	for i, tt := range tests {

--- a/vm/instruction.go
+++ b/vm/instruction.go
@@ -320,8 +320,7 @@ var builtInActions = map[operationType]*action{
 						t.returnError(InternalError, "Constant %s is not a class. got=%s", superClassName, string(superClass.Target.Class().ReturnName()))
 					}
 
-					class.pseudoSuperClass = inheritedClass
-					class.superClass = inheritedClass
+					class.inherits(inheritedClass)
 				}
 			}
 

--- a/vm/instruction.go
+++ b/vm/instruction.go
@@ -336,7 +336,7 @@ var builtInActions = map[operationType]*action{
 
 			switch r := receiver.(type) {
 			case *RClass:
-				if r.Singleton {
+				if r.isSingleton {
 					method = r.superClass.lookupInstanceMethod(methodName)
 				} else {
 					method = r.SingletonClass().lookupInstanceMethod(methodName)

--- a/vm/instruction.go
+++ b/vm/instruction.go
@@ -277,19 +277,7 @@ var builtInActions = map[operationType]*action{
 			method := &MethodObject{Name: methodName, argc: argCount, instructionSet: is, baseObj: &baseObj{class: t.vm.topLevelClass(methodClass)}}
 
 			v := t.stack.pop().Target
-
-			switch self := v.(type) {
-			case *RClass:
-				if self.pseudoSuperClass.Singleton {
-					self.pseudoSuperClass.ClassMethods.set(methodName, method)
-				}
-
-				class := t.vm.initializeClass(self.Name+"singleton", false)
-				class.Singleton = true
-				class.ClassMethods.set(methodName, method)
-				class.superClass = self.superClass
-				self.superClass = class
-			}
+			v.SingletonClass().Methods.set(methodName, method)
 			// TODO: Support something like:
 			// ```
 			// f = Foo.new
@@ -348,7 +336,12 @@ var builtInActions = map[operationType]*action{
 
 			switch r := receiver.(type) {
 			case *RClass:
-				method = r.lookupClassMethod(methodName)
+				if r.Singleton {
+					method = r.superClass.lookupInstanceMethod(methodName)
+				} else {
+					method = r.SingletonClass().lookupInstanceMethod(methodName)
+				}
+
 			default:
 				method = r.Class().lookupInstanceMethod(methodName)
 			}

--- a/vm/instruction.go
+++ b/vm/instruction.go
@@ -354,6 +354,10 @@ var builtInActions = map[operationType]*action{
 				method = r.Class().lookupInstanceMethod(methodName)
 			}
 
+			if receiver == t.vm.mainObj && methodName == "include" {
+				method, _ = t.vm.topLevelClass(classClass).Methods.get("include")
+			}
+
 			if method == nil {
 				t.UndefinedMethodError(methodName, receiver)
 				return

--- a/vm/object.go
+++ b/vm/object.go
@@ -7,6 +7,7 @@ import (
 // Object represents all objects in Goby, including Array, Integer or even Method and Error.
 type Object interface {
 	Class() *RClass
+	SingletonClass() *RClass
 	toString() string
 	toJSON() string
 	instanceVariableGet(string) (Object, bool)
@@ -31,6 +32,7 @@ type RObject struct {
 
 type baseObj struct {
 	class             *RClass
+	singletonClass    *RClass
 	InstanceVariables *environment
 }
 
@@ -40,6 +42,11 @@ func (b *baseObj) Class() *RClass {
 		panic(fmt.Sprint("Object doesn't have class."))
 	}
 	return b.class
+}
+
+// SingletonClass returns object's singleton class
+func (b *baseObj) SingletonClass() *RClass {
+	return b.singletonClass
 }
 
 func (b *baseObj) instanceVariableGet(name string) (Object, bool) {

